### ISSUE TITLE
Status Update (2019 - 2020)

### ISSUE THREAD

**Issue by ralfbiedert** [ralfbiedert/cheats.rs#100]
# Introduction

The site [**cheats.rs**](https://cheats.rs) was started because I was frustrated with the un-googleability of many constructs. I remember during my early Rust weeks I assumed `[0, 1]` and `[0; 1]` were the same thing, without even realizing there _could_ be a difference; `fn f(S { x }: S)` later was just _something they probably shot down over Roswell_.

The first iterations focused on a rather narrow interpretation of _language_ cheat sheet, contrasting sigils and keywords only. However, over time it became clear that condensing other _pragmatics_ into a few lines also made sense. Many of them followed my own learning and teaching trajectory, such as:

- Why can I pass around `&mut self` twice although _'&mut T references [are not] Copy_'?
- How should I design for closures in my API?
- What are the pitfalls of `async`-`await` again, and why?

With that said the project is in its second year now, and has been updated gradually over time. Where the Rust project uses editions to look back, I thought it would be nice to reflect a bit as well, what has actually happened the last year?

I also hope to get another round of critical reviews. While both content and traffic have increased over time, the amount of feedback (esp. Github issues) went down. A favorable interpretation could be that the writing has just gotten so-much-better, but more likely fewer people have taken the time to suggest improvements.

# Changes from 2019 to 2020

![a-b](https://user-images.githubusercontent.com/301288/89105554-5481f100-d422-11ea-96dd-4b03e772238e.png)

Some of the sections added the last 12 months:

### Behind the Scenes
- [**Memory & Lifetimes**](https://cheats.rs/#memory-lifetimes) (**total rewrite** of the previous version, just finished a few days ago, with much better visuals, I hope). If you only read one thing make sure it's this one.

### Data & Types
**Visual** section added a few months ago, explains most common types, constructs such as:

- [**Basic Types**](https://cheats.rs/#basic-types)
- [**Textual Types**](https://cheats.rs/#textual-types-ref)
- [**Custom Types**](https://cheats.rs/#custom-types)
- [**References & Pointers**](https://cheats.rs/#references-pointers-ui)
- [**Standard Library Types**](https://cheats.rs/#standard-library-types)

### Standard Library

- **Common One-Liners** (experimental)
- [**Traits**](https://cheats.rs/#traits) (tips and tricks around essential traits)
- [**String Conversions**](https://cheats.rs/#string-conversions) (recently updated to be more inference friendly)

### Tooling
- [**Project Anatomy**](https://cheats.rs/#project-anatomy) (incl. sample entry points)
- [**Cross Compilation**](https://cheats.rs/#cross-compilation) (probably the section I've sent out most links to)

### Coding Guides
- [**Async-Await 101**](https://cheats.rs/#async-await-101) (how to get started and what to avoid)
- [**Unsafe, Unsound, Undefined**](https://cheats.rs/#unsafe-unsound-undefined)
- [**API Stability**](https://cheats.rs/#api-stability) (condensed from an RFC that summarizes many things one should not do in stable APIs)

# Visuals

![visuals](https://user-images.githubusercontent.com/301288/89110035-e69bf080-d446-11ea-8cf7-4da2d6b3f7b2.png)

The majority of work went into creating visuals and there are now two major sections containing them:

- [**Data & Types**](https://cheats.rs/#data-types)
- [**Memory & Lifetimes**](https://cheats.rs/#memory-lifetimes)

The nice part about these visuals is that they are done entirely in HTML + CSS. Getting them to work in most browsers was mildly painful, but once done all fonts and colors became automatically consistent with the rest of the document. Also, they are easy to track in Git, and don't require tooling. They are also tested in 8 different configurations (Firefox, Chrome, Opera, Edge, iOS Tablet, iOS Phone portrait / landscape, PDF export) to make sure they work 'reasonably well' on most.

# Feedback System

![feedback](https://user-images.githubusercontent.com/301288/89105468-aece8200-d421-11ea-80b0-40c35a421876.png)

The site has a feedback system now. Every section should have 3 buttons when you hover over the title (still finalizing UX), that allow for immediate feedback on any platform, without the need to leave the page, let alone log into Github. I hope by reducing the friction from _saw an issue_ to _reported an issue_ review coverage and quality can be increased.

The feature is implemented as a simple JSON that includes either a `good`, `bad` or some text, along with meta data (e.g., the current target like `#memory-lifetimes`, or user agent for easier debugging). The backend is crude right now (experimental AWS Lambda written in their online code editor), but should store whatever is sent. Also, even when using the feedback system everything is still anonymous.

# Night Mode

![night](https://user-images.githubusercontent.com/301288/89105383-e688fa00-d420-11ea-9342-2088770b71ee.png)

There was a night mode for the longest time, but due to its implementation (pretty much a `filter: invert(1) hue-rotate(180deg);` on the `body`) it was one of these features only its developer could love, and even I didn't love it. With the latest update I took the time to rewrite all CSS and most colored tags, so that toggling night mode now does what it should: switching between two actual color themes.

# tl;dr

- New chapters
- New visuals
- Better night mode

[↗️ visit **cheats.rs**](https://cheats.rs), give feedback.
